File: kotlin/linked-list/src/main/kotlin/LinkedList.kt
class Deque<T> {

    data class Node<T>(val value: T, var next: Node<T>? = null, var previous: Node<T>? = null)

    private var head: Node<T>? = null
    private var tail: Node<T>? = null

    fun push(value: T) {
        val node = Node(value, previous = tail)

        if (tail == null) {
            head = node
        } else {
            tail?.next = node
        }

        tail = node
    }

    fun pop(): T? {
        val node = tail ?: throw NoSuchElementException()

        tail = node.previous

        if (tail == null) {
            head = null
        } else {
            tail?.next = null
        }

        return node.value
    }

    fun unshift(value: T) {
        val node = Node(value, next = head)

        if (head == null) {
            tail = node
        } else {
            head?.previous = node
        }

        head = node
    }

    fun shift(): T? {
        val node = head ?: throw NoSuchElementException()

        head = node.next

        if (head == null) {
            tail = null
        } else {
            head?.previous = null
        }

        return node.value
    }
}

File: csharp/linked-list/LinkedList.cs
public class Deque<T>
{
    class Node<U>(T value, Node<T>? next = null, Node<T>? previous = null)
    {
        public T Value { get; } = value;
        public Node<T>? Next { get; set; } = next;
        public Node<T>? Previous { get; set; } = previous;
    }

    private Node<T>? head = null;
    private Node<T>? tail = null;

    public void Push(T value)
    {
        var node = new Node<T>(value, null, tail);

        if (tail == null)
        {
            head = node;
        }
        else
        {
            tail.Next = node;
        }

        tail = node;
    }

    public T Pop()
    {
        var node = tail!;

        tail = node.Previous;

        if (tail == null)
        {
            head = null;
        }
        else
        {
            tail.Next = null;
        }

        return node.Value;
    }

    public void Unshift(T value)
    {
        var node = new Node<T>(value, head);

        if (head == null)
        {
            tail = node;
        }
        else
        {
            head.Previous = node;
        }

        head = node;
    }

    public T Shift()
    {
        var node = head!;

        head = node.Next;

        if (head == null)
        {
            tail = null;
        }
        else
        {
            head.Previous = null;
        }

        return node.Value;
    }
}